### ISSUE TITLE
Discuss this draft for changing how chain validations work

### ISSUE THREAD

**Issue by aesedepece** [witnet/witnet-rust#529]
## Introduction
The main idea of this proposal is:
- generalizing UTXOs and data request life cycles into a single data structure
- remove magic exceptions and validation rules on how UTXO set is managed (e.g. DR multispending)
- separating DR-related data needed for witnessing from data needed for block validations / chain syncing
- allowing stateful chain registers

## Key changes
This proposal introduces the concept of _Chain Register_ (tentative name):
- at any epoch, the entire state of the Witnet chain can be expressed as a set of chain registers
- chain registers are identified by their chain address, a.k.a. output pointers
- different types of chain registers have different properties
- all types of chain registers have a `value` property
- chain registers can be _stateless_ or _stateful_
- stateless registers are immutable, i.e. they can only be created and eventually dropped when they run out of value
- stateful registers are mutable, i.e. in addition to be created and eventually dropped, they can also be updated multiple times in between
- generally, different types of transactions can create, update or drop specific kinds of registers
- transaction inputs detract value from registers
- different register types allow different transaction types to detract different amounts of value (e.g. all value atomicly vs. a fixed sub-amount)
- outputs always create or update registers
- registers can only be dropped by inputs
- UTXOs are stateless registers whose value is atomically drained by transaction inputs, leading to their immediate dropping
- data requests are stateful registers whose value is partially consumed by commitment and reveal transactions. Tally transactions detract all the remaining value in the register and cause their eventual dropping
- further stateful register types can be created in the future to represent tokens subject to vesting or even tiny stateful smart contracts

This implies refactoring `UnspentOutputsPool` into `RegistersMap` and make it store all kind of registers, indexed by `RegisterAddress` (a.k.a. `OutputPointer`) and discriminated by type of register:
```rust
pub struct RegisterAddress {
    pub transaction_id: Hash,
    pub output_index: u32,
}

pub enum TaggedRegister {
  P2PKH(Output),
  Witnessing(WitnessingLifeCycle),
  DataPoint(RadonType),
  VestingToken(Vec<TokenInstallment>),
  Contract(ContractState)
}

pub type UnspentOutputsPool = HashMap<RegisterAddress, TaggedRegister>;
```

Different types of registers may implement their own `.spend()` methods to tell the spender whether the spending is allowed and how much value they can detract:
```rust
impl Register for Output {
  pub fn spend(&self, spender: &TypedTransaction) -> Result<u64, SpendingError> {
    if spender.proofs().valid_for(self) {
      Ok(self.value)
    } else {
      Err(SpendingError::InvalidSignature())
    }
  }
}

impl Register for WitnessingLifeCycle {
  pub fn spend(&self, spender: &TypedTransaction) -> Result<u64, SpendingError> {
    match spender {
      TypedTransaction::Commitment(spender) => {
        if spender.proofs().valid_for(self.query) {
          Ok(self.commitment_fee)
        } else {
          Err(SpendingError::InvalidPoE())
        }
      }
      /* ... */
    }
  }
}
```

#### Inputs
One or more pointers to `P2PKH` outputs.
##### Value
All value enclosed in the outputs pointed by the local inputs.
#### Outputs
One or more `P2PKH` outputs.
#### Proofs
As many `KeyedSignature`s as `P2PKH`s.
##### Validation logic
https://github.com/witnet/witnet-rust/issues/526#issuecomment-474445280
#### Miner fees
Implicit by `SUM(inputs) - SUM(outputs)`.
#### Registers actions
- Drain and drop all registers pointed by inputs.
- Insert outputs as `P2PKH` registers.
#### Specific validation rules
None.

#### Inputs
One or more pointers to `P2PKH` outputs.
##### Value
All value enclosed in the outputs pointed by the local inputs.
#### Outputs
One or more `DataRequest` outputs.
Optionally, one or more `P2PKH` outputs (e.g. for "change")
#### Proofs
As many `KeyedSignature`s as `P2PKH`s.
##### Validation logic
- https://github.com/witnet/witnet-rust/issues/526#issuecomment-474445280
#### Miner fees
Implicit by `SUM(inputs) - SUM(outputs)`.
#### Registers actions
- Darin and drop pointed `P2PKH` registers.
- Insert `DataRequest` outputs as `Witnessing` registers.
- Insert `P2PKH` outputs as `P2PKH` registers.
#### Specific validation rules
Imperatively, none.
Optionally, semantic validity of data request body.

#### Inputs
One and only one pointer to a `DataRequest` output.
Optionally, one or more pointers to spendable `P2PKH` to use as collateral / sybil deterrents.
##### Value
- The explicit `commit_fee` from the pointed `DataRequest`, plus
- All value enclosed in the outputs pointed by any local `P2PKH` inputs.
#### Outputs
One and only one `Commitment` output.
#### Proofs
1. One and only one `KeyedProofOfEligibility`.
2. As many `KeyedSignature`s as `P2PKH`s.
##### Validation logic
- `DataRequest`: `KeyedProofOfEligibility` must be valid for the pointed `DataRequest`.
- `P2PKH`: https://github.com/witnet/witnet-rust/issues/526#issuecomment-474445280
#### Miner fees
- Only the explicit `commitment_fee` from the pointed `DataRequest`
#### Registers actions
- Drain and drop all pointed `P2PKH` registers.
- Update `WitnessingLifeCycle` by pushing itself into the `commitments` vector as `TaggedTransaction::Commitment` (extracting the output and tagging it with the collateral value should work too)
#### Specific validation rules
- `P2PKH`s must be older than `MINIMUM_COLLATERAL_AGE`.
- Value brought by `P2PKH`s must exceed `MINIMUM_COLLATERAL_VALUE` plus `collateral` from the pointed `Witnessing` register.

#### Inputs
One and only one pointer to a `DataRequest` output.
##### Value
- The explicit `commit_fee` from the pointed `DataRequest`
#### Outputs
One and only one `Reveal` output.
#### Proofs
One and only one `NoncedReveal`.
##### Validation logic
`NoncedReveal` must be valid for the pointed `DataRequest`.
#### Miner fees
- Only the explicit `reveal_fee` from the pointed `DataRequest`
#### Registers set actions
- Update `WitnessingLifeCycle` by pushing itself into the `reveals` vector as `TaggedTransaction::Reveal` (extracting the output should work too)
#### Specific validation rules
None

#### Inputs
One and only one pointer to a `DataRequest` output.
##### Value
- The explicit `tally_fee` from the pointed `DataRequest`
#### Outputs
- One and only one `Tally` output.
- As many `P2PKH` outputs as reveals have been found to agree with consensus. The value of each `P2PKH` must equal `reward + collateral` from the pointed `DataRequest`.
#### Proofs
None.
##### Validation logic
None.
#### Miner fees
- Only the explicit `tally_fee` from the pointed `DataRequest`
#### Registers set actions
- Replace `WitnessingLifeCycle` register by a new `DataPoint` register containing the tally, exactly at same address.
#### Specific validation rules
`Tally` value must match the result of applying the `consensus_script` from `DataRequest` on the `reveals` array from `DataRequest`.

**Comment by aesedepece** [comment]
Related: https://github.com/witnet/witnet-rust/pull/538#discussion_r269534527

**Comment by aesedepece** [comment]
The team considered this proposal positive, necessary and opportune. This issue will be closed and work will continue on #557.
